pkg/controllers/machinesetsync/machineset_sync_controller.go

```diff
@@ -415,3 +577,41 @@ func objectMetaIsEqual(a, b metav1.ObjectMeta) bool {
 		reflect.DeepEqual(a.Finalizers, b.Finalizers) &&
 		reflect.DeepEqual(a.OwnerReferences, b.OwnerReferences)
 }
+
+// capiInfraMachineTemplateIsEqual checks whether the provided CAPI infra machine templates are equal.
+func capiInfraMachineTemplateIsEqual(platform configv1.PlatformType, infraMachineTemplate1, infraMachineTemplate2 client.Object) (bool, error) {
+	switch platform {
+	case configv1.AWSPlatformType:
+		typedInfraMachineTemplate1, ok := infraMachineTemplate1.(*awscapiv1beta1.AWSMachineTemplate)
+		if !ok {
+			return false, errAssertingCAPIAWSMachineTemplate
+		}
+
+		typedinfraMachineTemplate2, ok := infraMachineTemplate2.(*awscapiv1beta1.AWSMachineTemplate)
+		if !ok {
+			return false, errAssertingCAPIAWSMachineTemplate
+		}
+
+		return reflect.DeepEqual(typedInfraMachineTemplate1.Spec, typedinfraMachineTemplate2.Spec) && objectMetaIsEqual(typedInfraMachineTemplate1.ObjectMeta, typedinfraMachineTemplate2.ObjectMeta), nil
+	default:
+		return false, fmt.Errorf("%w: %s", errPlatformNotSupported, platform)
+	}
+}
+
+// getResourceVersion returns the object ResourceVersion or the zero value for it.
+func getResourceVersion(obj client.Object) string {
+	if obj == nil {
+		return "0"
+	}
+
+	return obj.GetResourceVersion()
+}
+
+// getCAPIMSResourceVersion returns the CAPI machine set ResourceVersion or the zero value for it.
+func getCAPIMSResourceVersion(obj *capiv1beta1.MachineSet) string {
+	if obj == nil {
+		return "0"
+	}
+
+	return obj.GetResourceVersion()
+}
```

Comment:
Why is this called out separately, the `getResourceVersion` should function the same no?

Reply:
That requires casting *capiv1beta1.MachineSet to client.Object() before passing that into getResourceVersion and also using reflect.ValueOf to check the inner value of the interface-wrapped type.
I had a go at it and it seems to work 👍